### PR TITLE
Fix unit tests in QuestionnaireViewModelTest

### LINKED ISSUE #1820: Tests in QuestionnaireViewModel pass when they should pass
Our `QuestionnaireViewModelTest` is passing tests that should be failing (the worst kind). 

The code is making a cast that should never succeed:
```
viewModel.questionnaireStateFlow.value as DisplayMode.EditMode
```

but that is being called inside viewModel.runViewModelBlocking, which seems to run after the test completes. The custom function runs on the main dispatcher, but that does not block the test from completing.

https://github.com/google/android-fhir/blob/master/datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt#L4043-L4057

### PR DESCRIPTION
**Pull request by omarismail94**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #1820 

**Description**
runViewModelBlocking function was swallowing incorrect casts and failures. This is fixed now. 

/**
 * In local unit tests, the Main dispatcher that wraps the Android UI thread will be unavailable, as
 * these tests are executed on a local JVM and not an Android device.
 * [androidx.lifecycle.viewModelScope], which we use in [QuestionnaireViewModel], uses a hardcoded
 * Main dispatcher under the hood, which needs to be replaced with a TestDispatcher.
 *
 * See: https://developer.android.com/kotlin/coroutines/test#setting-main-dispatcher
 *
 * The TestDispatcher we create is then used to launch a job to collect the results from
 * [QuestionnaireViewModel.questionnaireStateFlow]
 *
 * See: https://developer.android.com/kotlin/flow/test#statein
 */

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

**Type**
Choose one: (Bug fix | Feature | Documentation | Testing | Code health | Builds | Releases | Other)

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

**Comment by omarismail94** [comment]
TODO: get the two tests that are failing to pass

**Review by aditya-07 — COMMENTED** [review]
We should also add a small writeup describing how to write these tests with turbine for anyone adding tests in the future.

File: buildSrc/src/main/kotlin/JacocoConfig.kt [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2021 Google LLC
+ * Copyright 2022 Google LLC
```
Comment (jingtang10) [inline buildSrc/src/main/kotlin/JacocoConfig.kt:2]: please revert

File: datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt [hunk 2]
```diff
@@ -3997,10 +3994,10 @@ class QuestionnaireViewModelTest {
       "questionnaireResponseItem"
     )
 
-  private companion object {
+  internal companion object {
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt:3997]: revert

File: datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt [hunk 3]
```diff
@@ -3997,10 +3994,10 @@ class QuestionnaireViewModelTest {
...
     const val CODE_SYSTEM_YES_NO = "http://terminology.hl7.org/CodeSystem/v2-0136"
 
-    private val paginationExtension =
+    internal val paginationExtension =
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt:4000]: revert

File: datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt [hunk 4]
```diff
@@ -4013,7 +4010,7 @@ class QuestionnaireViewModelTest {
         )
       }
 
-    private val hiddenExtension =
+    internal val hiddenExtension =
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt:4013]: revert

File: datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt [hunk 5]
```diff
@@ -1113,212 +1114,209 @@ class QuestionnaireViewModelTest {
...
+        assertThat(items.map { it.questionnaireItem.linkId })
+          .containsExactly("question-1", "question-2", "question-3")
 
-      items = viewModel.getQuestionnaireItemViewItemList().map { it.asQuestion() }
-      assertThat(items.map { it.questionnaireItem.linkId }).containsExactly("question-1")
+        items.first { it.questionnaireItem.linkId == "question-1" }.clearAnswer()
+      }
+
+      viewModel.questionnaireStateFlow.test {
+        assertThat(awaitItem().items.map { it.asQuestion() }.map { it.questionnaireItem.linkId })
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt:1203]: as discussed:
1. try if you can call awaitItem multiple times to get to the later state
2. see what you can do with `getQuestionnaireItemViewItemList`

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt [hunk 6]
```diff
@@ -316,7 +316,11 @@ internal class QuestionnaireViewModel(application: Application, state: SavedStat
 
         val allCurrentPageItemsValid =
           currentPageItems.filterIsInstance<QuestionnaireAdapterItem.Question>().all {
-            it.item.validationResult is Valid
+            QuestionnaireResponseItemValidator.validate(
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:319]: can you explain why this is needed?

**Review by jingtang10 — CHANGES_REQUESTED** [review]
thanks so much @omarismail94!

File: datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt [hunk 7]
```diff
@@ -1113,212 +1114,209 @@ class QuestionnaireViewModelTest {
...
+        .setAnswer(
           QuestionnaireResponse.QuestionnaireResponseItemAnswerComponent().apply {
             value = BooleanType(true)
           }
         )
 
+      assertThat(
+          (questionOne.getQuestionnaireResponseItem().answer[0].value as BooleanType).booleanValue()
+        )
+        .isEqualTo(true)
```
[the remark this answers is not in the document]
Comment (omarismail94) [reply datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt]: I got rid of Turbine and found a way to get the test to work. We need to add our own dispatcher to override the hardcoded viewModelScope, and make sure we launch any collect jobs in it as well. Please see the documentation I added to the top of the file

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt (hunk 6 above)
Comment (omarismail94) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt]: we dont need this anymore. I found a way to run the tests correctly. The issue was the tests were not running in the viewModelScope. I fixed this!  Please see the documentation I added to the top of the file that links to how I got to this solution!

File: datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt (hunk 5 above)
Comment (omarismail94) [reply datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt]: This is outdated as I got rid of the need to have turbine.awaitItem

File: datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt [hunk 8]
```diff
@@ -3434,20 +3476,22 @@ class QuestionnaireViewModelTest {
...
+      viewModel.goToNextPage()
+      val collectJob =
+        launch(mainDispatcherRule.testDispatcher) { viewModel.questionnaireStateFlow.collect() }
+
+      assertThat(
+          (viewModel.questionnaireStateFlow.value.displayMode as DisplayMode.EditMode)
+            .pagination.showReviewButton
+        )
+        .isFalse()
+      collectJob.cancel()
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt:3489]: you could have a private helper function:

```
private fun something(block: {}) {
  val collectJob = ...
  block.run()
  collectJob.canel()
}
```

**Review by jingtang10 — APPROVED** [review]

File: datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt (hunk 8 above)
Comment (omarismail94) [reply datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt]: Implemented this! Great suggestion as this cleaned up the code. I also added a test case where we do an incorrect cast just to check that an exception is raised and is not swallowed like before

**Pull request merged**